src/main/java/seedu/address/model/recipe/attribute/IngredientList.java

```diff
@@ -1,48 +1,29 @@
 package seedu.address.model.recipe.attribute;
 
-import static java.util.Objects.requireNonNull;
-import static seedu.address.commons.util.AppUtil.checkArgument;
+import static seedu.address.commons.util.CollectionUtil.requireAllNonNull;
 
 import java.util.ArrayList;
 import java.util.List;
+import java.util.stream.Collectors;
 
 /**
  * Represents a Recipe's ingredients in the recipe book.
  */
 public class IngredientList {
-
-    public static final String MESSAGE_CONSTRAINTS =
-            "Ingredients can take any values for now, and it should not be " + "blank";
-
-    /*
-     * The first character of the ingredients must not be a whitespace,
-     * otherwise " " (a blank string) becomes a valid input.
-     */
-    public static final String VALIDATION_REGEX = "[^\\s].*";
-
-    public final List<Ingredient> ingredients;
-
-    // TODO: change to List, for now using String to temp store ingredient list
-    public final String ingredientListString;
+    private final ArrayList<Ingredient> ingredientData;
 
     /**
-     * Constructs an {@code IngredientList}.
+     * Constructs an ingredient list from {@code ingredientList}. Requires that the
+     * parameter be non-null. The parameter is guaranteed to be <em>valid</em>, as
+     * every item in the list has already been individually checked by
+     * {@link Ingredient#Ingredient(String)}.
+     *
+     * @param ingredientList a {@link List} of {@link Ingredient}s
      */
-    public IngredientList(String ingredientListString) {
-        requireNonNull(ingredientListString);
-        checkArgument(isValidIngredients(ingredientListString), MESSAGE_CONSTRAINTS);
-
-        this.ingredientListString = ingredientListString;
-
-        // TODO: Update IngredientList to use arraylist instead of raw String
-        this.ingredients = new ArrayList<Ingredient>();
-    }
+    public IngredientList(List<Ingredient> ingredientList) {
```

Comment:
Each ingredient in the list is guaranteed to be non-null and valid, as mentioned in the header comments. No need for an additional `checkArgument(...)` here, I feel.